Change indexing of work and work versions

awead opened the pull request:
Both public and dashboard search interfaces now return only Work resources for search results. All works are now indexed, regardless of what kinds of versions they have. Any given work will have a "state" that is determined by its versions:

  * published work = work with any published version
  * withdrawn work = work with any withdrawn versions and no published
    versions
  * draft work = work with no published or withdrawn versions

All the relevant search builder models were updated to reflect this, as well as changes to work policies so that draft and 
withdrawn works will resolve for users who have access to them.

Fixes #1026

cam156 commented on app/services/user_stats_importer.rb:
```diff
@@ -118,4 +79,17 @@ def create_or_update_object_stat(event, user_db_id, object, stat_class)
       updated_stat.save
     end
+
+    def create_or_update_user_stats(stats, user)
```
Should we add a comment that this code was copied from Sufia::UserStat?

cam156 commented:
@awead I'm not certain if this is an accurate code coverage drop.  Are you happy with the coverage on this PR?

awead commented:
@cam156 yeah, I wasn't either. I'm happy if you are!

cam156 commented:
@awead if I am reading code coverage right it looks like import is not being tested
https://coveralls.io/builds/13847428/source?filename=app%2Fservices%2Fuser_stats_importer.rb#L12

awead committed:
Change indexing of work and work versions

Both public and dashboard search interfaces now return only Work
resources for search results. All works are now indexed, regardless of
what kinds of versions they have. Any given work will have a "state"
that is determined by its versions:

  * published work = work with any published version
  * withdrawn work = work with any withdrawn versions and no published
    versions
  * draft work = work with no published or withdrawn versions

All the relevant search builder models were updated to reflect this, as
well as changes to work policies so that draft and withdrawn works will
resolve for users who have access to them.
